### DISCUSSION TITLE
Compiling source

### DISCUSSION THREAD

**Discussion by AmirrezaHBSH** [firemodels/fds#12690] [Q&A]
Hello, I downloaded the FDS source to my cluster using the command: git clone. However, I encountered an issue while compiling the code. The error indicates the presence of ^M characters in the code, which need to be removed for successful compilation. I'm seeking advice on alternative or improved methods for downloading the code to the cluster to prevent encountering this error in the future.

**Comment by mcgratta** [comment]
Our default and preferred text file format is unix/linux. Can you tell us which file is causing the issue?

**Reply by AmirrezaHBSH** [reply]
I am not sure which file is causing the issue, since I am a new user. In the Fire models when I want to compile fds file, I face with this issue. I am using PuTTY.

**Comment by gforney** [comment]
by any chance did you clone the fds repo on a Windows PC then copy the files to a Linux or Mac ?  if so you need to use the git clone command on the computer where the files will reside

**Reply by AmirrezaHBSH** [reply]
No, I just downloaded the fds repo directly to the cluster by using PuTTY

**Comment by mcgratta** [comment]
I see that imkl.f90 looks MacOS-y. I'll look for others.

**Comment by mcgratta** [comment]
My vi editor tells me this
```
"imkl.f90" [noeol] 197L, 9271C
```
But this has never caused us any trouble.

**Comment by mcgratta** [comment]
I changed [one file](https://github.com/firemodels/fds/pull/12691/files) but I doubt that this is the cause of your issue. You could update the repo and try again.

**Reply by AmirrezaHBSH** [reply]
How can I download just that edited file in my cluster?

**Reply by TristanHehnen** [reply]
On the Linux machine, navigate to your git repo and execute `git pull`.

**Comment by mcgratta** [comment]
I have never used PuTTy to clone a git repository. How did you do it?

**Reply by AmirrezaHBSH** [reply]
I used this command: git clone git@github.com:firemodels/fds.git

**Reply by TristanHehnen** [reply]
There may be a misunderstanding here. 
Putty is used to enable an ssh connection from Windows machines to Linux. One would end up with a regular Linux terminal on the remote machine.

https://hub.wpi.edu/article/785/connect-to-linux-cluster-on-windows-using-putty


Thus, the git clone should be a regular terminal interaction and nothing specific to putty itself.

**Comment by mcgratta** [comment]
OK, thanks. So assuming that the fds repository has been cloned directly onto your linux-based cluster. How exactly are you trying to compile the code?

**Reply by AmirrezaHBSH** [reply]
I am using ompi-gnu-linux, and make fds.sh. In this step I can see an error of ^M. then I am removing this character then compiling.

**Comment by mcgratta** [comment]
Do you know what file has the ^M character?

**Reply by AmirrezaHBSH** [reply]
:ompi_gnu_linux> ./make_fds.sh
-bash: ./make_fds.sh: /bin/bash^M: bad interpreter: No such file or directory

This is the error I have when compiling fds all files. Could you please tell me how I can search for the exact file to see which file is causing this issue?

**Comment by mcgratta** [comment]
This appears to be a problem with the file `make_fds.sh`. Open this file with the `vi` editor and you should see this
```
#!/bin/bash
dir=`pwd`
target=${dir##*/}

echo Building $target
make -j4 VPATH="../../Source" -f ../makefile $target
~
~
"make_fds.sh" 6L, 116C                                     
```
Make sure that is does not say `DOS` at the bottom. Then at the command prompt, type the command `ls -l`
```
ls -l
total 8
-rwxr-xr-x. 1 mcgratta div733 116 Nov 29  2022 make_fds.sh
```
Do you see the same arrangement of r's and x's indicating that the file is an executable script? If not
```
chmod +x make_fds.sh
```

**Reply by AmirrezaHBSH** [reply]
Thank you for your reply, however, vi editor does not show anything inside the make_fds.sh

**Reply by AmirrezaHBSH** [reply]
Yes I got it there is DOS at the bottom. How should I procced?

**Reply by mcgratta** [reply]
Quick solution would be to change the file `make_fds.sh` to unix format. I do this by opening the file with `vi` and then type
```
set ff=unix
```
Then `wq` to exit.  Longer term, you should figure out why your cloning of the repository changed the line endings.

**Reply by AmirrezaHBSH** [reply]
Where should I type this set ff=unix? at the beginning of the vi editor?

**Comment by mcgratta** [comment]
```
:set ff=unix
```
This is a command issued by the editor. The colon gives you a prompt

**Reply by AmirrezaHBSH** [reply]
I appreciate. I did that got this error:  

Building ompi_gnu_linux
/bin/bash: ../Scripts/osx_clt.sh: /bin/bash^M: bad interpreter: No such file or directory
mpifort -c -m64 -O2 -std=f2018 -frecursive -ffpe-summary=none -fall-intrinsics -cpp -DGITHASH_PP=\"FDS-6.9.0-13-gd1b1af3-master\" -DGITDATE_PP=\""Wed Mar 27 09:36:17 2024 -0400\"" -DBUILDDATE_PP=\""Mar 27, 2024  10:05:34\""   ../../Source/prec.f90
mpifort -c -m64 -O2 -std=f2018 -frecursive -ffpe-summary=none -fall-intrinsics -cpp -DGITHASH_PP=\"FDS-6.9.0-13-gd1b1af3-master\" -DGITDATE_PP=\""Wed Mar 27 09:36:17 2024 -0400\"" -DBUILDDATE_PP=\""Mar 27, 2024  10:05:34\""   ../../Source/imkl.f90
/bin/bash: mpifort: command not found
make: *** [../makefile:121: prec.o] Error 127
make: *** Waiting for unfinished jobs....
/bin/bash: mpifort: command not found
make: *** [../makefile:121: imkl.o] Error 127

**Reply by marcosvanella** [reply]
You have to have openmpi loaded in your environment to have access to mpifort. Ask your cluster IT person.
Also do the same change to unix on the script in Build/Scripts/osx_clt.sh. Somehow you got these ^M DOS end of line characters when cloning the repo.

**Comment by mcgratta** [comment]
You need IT support for the particular computer you are using. We cannot help you solve issues related to the installation of software on your computer.
